### DISCUSSION TITLE
NextJS Internationalized Routing - Get user locale

### DISCUSSION THREAD

**Discussion by M1ck0** [vercel/next.js#20085] [Help] [answered]
> When a user visits the application root (generally /), Next.js will try to automatically detect which locale the user prefers based on the Accept-Language header and the current domain.

As we know this is a current strategy that `Next` uses for getting the user's default locale.

Is it possible to change which header does `Next` read?

The thing is that we need to get the user's locale based on his IP address (location), which will be done on the server backend and sent via a custom header.

Is it possible to make `Next` read that custom header instead of `Accept-Language`, or do we need to install additional libraries to make this work?

Thank you for your time

**Comment by M1ck0** [comment]
Bump?

**Comment by M1ck0** [comment]
Anyone?

**Comment by stigkj** [comment, marked as answer]
As far as I know, it is the Next.js server backend that reads the `accept-language` header. This means the language is already decided when your code is executed on the server. That is if you are not talking about some kind of proxy in front of your Next.js app.

Next.js has support for overriding the `accept-language` header with a [cookie](https://nextjs.org/docs/advanced-features/i18n-routing#leveraging-the-next_locale-cookie), i.e. you could find the user's locale based on the IP address, set the `NEXT_LOCALE` cookie in your code, and force a page reload so the correct language is loaded.

**Reply by JustFly1984** [reply]
What about the cases for countries or cities with multilingual population? What if user switches language at runtime? How to keep preferred language on the backend? Currently I keep preferred language on the client in local storage, and redirect to correct page at runtime

**Reply by stigkj** [reply]
@JustFly1984 @M1ck0 For switching language you can use the built-in support via the cookie, see https://nextjs.org/docs/advanced-features/i18n-routing#leveraging-the-next_locale-cookie.

tl;dr: when changing language, set the cookie and redirect to the correct language with `<Link href="/this_page" locale="fr"/>` or `router.push('/this_page', '/this_page', { locale: 'fr' })`

**Reply by M1ck0** [reply]
@stigkj `NEXT_LOCALE` cookie is not working as intended. See #22863, but I guess no one is interested in this.
